Adding support for 126 more fiat currencies

robbiehanson opened the pull request:
Added nearly every fiat currency listed [here](https://oxr.readme.io/docs/supported-currencies).

<img height="600" alt="Screen Shot 2021-10-26 at 11 12 05" src="https://user-images.githubusercontent.com/304604/138918955-e1893238-da08-4c23-af02-11a224b242bc.png">


Major architectural changes:

Previously we had one type `BitcoinPriceRate`, which converts between BTC & Fiat. We now have an additional type `UsdPriceRate` which converts between USD & Fiat. When combined with a `BitcoinPriceRate` for USD, we have a reliable conversion system for any currency, even when a direct BTC-FIAT rate isn't available, or the corresponding market isn't sufficiently deep/liquid.

Corresponding to this change, we've modified the database to support storing `UsdPriceRate` values.

Additional logic has been added to support different refresh rates for BTC-FIAT vs USD-FIAT, since the USD-FIAT updates at most once per-hour (as per source).

Notes:

I've tried to remove currencies which are no longer in active use. For example, the currencies listed as "in use prior to 2016".  But I'm no expert in this field, so there may be more.

robbiehanson commented on phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/CurrencyManager.kt:
```diff
@@ -53,36 +57,96 @@ class CurrencyManager(
+        // That's an average wait time of 23 seconds before a given fiat currency is ready.
+        // We're attempting to maximize concurrency by fetching all,
+        // and hopefully using http pipelining to speed up the process.
+        // The end result is that the process takes around 10 seconds.
```
These are good points worth investigating before we move forward with this PR.



> I worry that iOS may somehow be annoyed if Phoenix executes 150+ http GET queries when the app runs in the background and network access is restricted.



Theoretically, on iOS, ktor uses Apple's URLSession, which automatically manages connections. Which means that if you issue multiple GET requests to the same host, then Apple may only create a single TCP/TLS connection. However, that's theoretical. I would like to verify this is the case with the CoinDesk API. (e.g. they're not force-closing the connection after each response)



If the internet isn't available, then the `AppConnectionsDaemon` calls `currencyManager.stop()`. But I need to verify that it won't call `currencyManager.start()` when the app is running in the background.



> Also logs could get spammed by `error` lines if the coindesk API is not reachable which can happen quite often.



Good point. I will investigate.



> the shared `CurrencyManager` would need to be aware of the user preferences, which is not the case now



This is a good idea. Even if we still want to fetch every rate, we should prioritize the user's preference.

robbiehanson replied:
It turns out that ktor has a bug, which results in significant performance degradation on iOS. At least for this particular use case.

https://youtrack.jetbrains.com/issue/KTOR-3362

Here's the cliff notes:
- if you use the same NSURLSession for each request, then Apple automatically creates a single TCP/TLS connection. And then shares the same HTTP/2 connection for each request. And in testing, all responses complete in 800-1400 milliseconds.
- but ktor instead creates a new NSURLSession for each request (see code [here](https://github.com/ktorio/ktor/blob/main/ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt#L52))
- thus we don't get to take advantage of **any** of the optimizations Apple worked so hard to create. And instead, a new TCP/TLS connection is created for every single request. And in testing, all responses complete in 6-7 seconds.

robbiehanson replied:
- [X] CurrencyManager refresh task are cancelled when app moves into the background (on iOS)
  - This is accomplished via `AppConnectionsDeamon` which invokes `currencyManager.stop()`
- [X] CurrencyManager won't run when app is in background, and processing a push notification
  - Fixed in 59162f2

dpad85 replied:
With the `preferred` parameter from 94a829d in `refreshFromCoinDesk` we'll be able to easily tweak the manager if need be 👍

robbiehanson commented on phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/CurrencyManager.kt:
```diff
@@ -40,49 +65,178 @@ class CurrencyManager(
+        val fiatCurrencies = FiatCurrency.values.filter {
+            it != FiatCurrency.USD && it != FiatCurrency.EUR
+        }
+        return fiatCurrencies.map { fiatCurrency ->
```
It's like that because I needed the `fiatCurrencies` variable again below:

```kotlin
appDb.saveExchangeRates(fetchedRates)
log.info { "successfully refreshed exchange rates from coindesk.com" }
// We consider the fetch to be successful if we fetched at least 2/3 of the list
fetchedRates.size >= (fiatCurrencies.size * 2 / 3) // ^let
```

However, I'm not particularly happy about this fuzzy logic...

I know it's possible that one currency starts failing, possibly because CoinDesk drops support for it. And I didn't want this possibility to cause excessive refreshes of ALL the currencies. But let me see if I can rework this...

robbiehanson replied:
Fixed in 29196b0

robbiehanson commented on phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/CurrencyManager.kt:
```diff
@@ -40,49 +65,178 @@ class CurrencyManager(
+                // This means that:
+                // - 1 USD = 3.2082 ILS
+                // - 1 ILS = 62,980.0572 * 2.2082 BTC = 202,056.4047 BTC
+                val price = fiatRate.rate / usdRate.rate
```
Good question. And I originally started coding it this way, but changed it for 2 reasons:

First, we know that CoinDesk is using the USD/BTC + ILS/USD to generate the ILS/BTC rate. But we also know that it only updates the fiat exchange rates (ILS/USD) once per hour. So if we had the raw ILS/USD rate, then we could reduce how often we fetch the fiat exchange rates from the API.

That is, it allows us to use independent rates for fetching USD/BTC vs ILS/USD. Currently we refresh USD/BTC every 20 minutes, and ILS/USD every hour.

Second, it provides a level of independence from the CoinDesk API. If the CoinDesk API disappears, we can replace it with any other fiat/fiat conversion API. 

I suspect we may consider doing so anyway. We know that CoinDesk is pulling its fiat exchange rates from [openexchangerates.org](https://openexchangerates.org/). And if we go direct to this API, we could also start asking for historical exchange rates. (e.g. what was the USD/ILS rate on 2021-10-14 at 14:02) And historical exchange rates is one of the frequent feature requests we have for Phoenix. (e.g. "I don't want to see the current price of 15,000 sat. I want to see how much money that was when I actually made the transaction 14 months ago...")

dpad85 replied:
Got it, thanks for the in-depth explanation.

robbiehanson commented:
> If that's fine with you I'll push it to this PR.

Looks good :thumbsup:

robbiehanson commented:
I've refactored the refresh logic. Here's what's new:



We now perform exponential backoff. And the exponential backoff logic is per currency. So if a particular currency fails (or a subset fails), then we only retry those that failed. To accomplish this we now keep a `refreshList` with information about each currency.



We keep track of the "preferred" fiat currencies (according to the app), and we prioritize those by fetching them first.



I refactored the logic to avoid spamming the logs if we get a series of failures. On either success or failure, we're only going to get a single output in the logs.

dpad85 reviewed (approved):
🚀
